File: desktop/windows/Nexus.Desktop/Sidecar/SidecarSupervisor.cs
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using Nexus.Desktop.Diagnostics;
using Nexus.Desktop.Runtime;

namespace Nexus.Desktop.Sidecar;

internal sealed class SidecarSupervisor : IDisposable
{
    private readonly DesktopStartupTimeline startupTimeline;
    private readonly SidecarBundle locator;
    private readonly SidecarRuntimeConfig runtime;
    private Process? process;

    public SidecarSupervisor(DesktopStartupTimeline startupTimeline)
    {
        this.startupTimeline = startupTimeline;
        locator = SidecarBundleLocator.Resolve();
        int port = SidecarPortAllocator.Allocate();
        runtime = new SidecarRuntimeConfig(
            Port: port,
            SessionToken: DesktopSessionToken.Generate(),
            AppVersion: AppVersionInfo.Version,
            BuildNumber: AppVersionInfo.BuildNumber,
            Platform: "windows");
        startupTimeline.Mark("sidecar.config_resolved", new Dictionary<string, string>
        {
            ["mode"] = locator.IsDevelopment ? "development" : "bundle",
            ["port"] = port.ToString(),
        });
    }

    public async Task<SidecarRuntimeConfig> StartAsync()
    {
        startupTimeline.Mark("sidecar.launch_begin");
        ProcessStartInfo startInfo = BuildStartInfo();
        process = Process.Start(startInfo) ?? throw new InvalidOperationException("无法启动 nexus-server。");
        process.OutputDataReceived += (_, args) =>
        {
            if (!string.IsNullOrWhiteSpace(args.Data))
            {
                Trace.WriteLine($"[Nexus Sidecar stdout] {args.Data}");
            }
        };
        process.ErrorDataReceived += (_, args) =>
        {
            if (!string.IsNullOrWhiteSpace(args.Data))
            {
                Trace.WriteLine($"[Nexus Sidecar stderr] {args.Data}");
            }
        };
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();
        startupTimeline.Mark("sidecar.process_started", new Dictionary<string, string>
        {
            ["pid"] = process.Id.ToString(),
        });

        await WaitUntilHealthyAsync();
        startupTimeline.Mark("sidecar.health_ready");
        return runtime;
    }

    public void Dispose()
    {
        if (process is { HasExited: false })
        {
            process.Kill(entireProcessTree: true);
            process.WaitForExit(3000);
        }
        process?.Dispose();
    }

    private ProcessStartInfo BuildStartInfo()
    {
        PrepareDirectories();
        DesktopCredentialsKey credentialsKey = DesktopCredentialsKeyStore.ConnectorCredentialsKey();
        startupTimeline.Mark("sidecar.credentials_key_ready", new Dictionary<string, string>
        {
            ["storage"] = credentialsKey.Storage,
            ["reason"] = credentialsKey.Reason,
        });

        var startInfo = new ProcessStartInfo
        {
            FileName = locator.Command,
            Arguments = locator.Arguments,
            WorkingDirectory = locator.WorkingDirectory,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
        };

        startInfo.Environment["NEXUS_APP_MODE"] = "desktop";
        startInfo.Environment["NEXUS_APP_ROOT"] = locator.AppRoot;
        startInfo.Environment["NEXUS_CONFIG_DIR"] = DesktopPaths.ConfigDirectory;
        startInfo.Environment["HOST"] = "127.0.0.1";
        startInfo.Environment["PORT"] = runtime.Port.ToString();
        startInfo.Environment["NEXUS_DESKTOP_SESSION_TOKEN"] = runtime.SessionToken;
        startInfo.Environment["WEB_DIST_DIR"] = locator.WebDistDirectory;
        startInfo.Environment["DATABASE_DRIVER"] = "sqlite";
        startInfo.Environment["DATABASE_URL"] = Path.Combine(DesktopPaths.ApplicationDataDirectory, "nexus.db");
        startInfo.Environment["CONNECTOR_CREDENTIALS_KEY"] = credentialsKey.Value;
        startInfo.Environment["WORKSPACE_PATH"] = DesktopPaths.WorkspaceDirectory;
        startInfo.Environment["CACHE_FILE_DIR"] = DesktopPaths.CacheDirectory;
        startInfo.Environment["LOG_PATH"] = Path.Combine(DesktopPaths.LogsDirectory, "sidecar.log");
        startInfo.Environment["LOG_STDOUT"] = "true";
        startInfo.Environment["LOG_FILE_ENABLED"] = "true";
        startInfo.Environment["DISCORD_ENABLED"] = "false";
        startInfo.Environment["TELEGRAM_ENABLED"] = "false";
        startInfo.Environment["CONNECTOR_OAUTH_REDIRECT_URI"] = "nexus://connectors/oauth/callback";
        startInfo.Environment["CONNECTOR_OAUTH_ALLOWED_ORIGINS"] = $"{runtime.WebBaseUrl.TrimEnd('/')},nexus://connectors";
        return startInfo;
    }

    private static void PrepareDirectories()
    {
        Directory.CreateDirectory(DesktopPaths.ApplicationDataDirectory);
        Directory.CreateDirectory(DesktopPaths.ConfigDirectory);
        Directory.CreateDirectory(DesktopPaths.WorkspaceDirectory);
        Directory.CreateDirectory(DesktopPaths.CacheDirectory);
        Directory.CreateDirectory(DesktopPaths.LogsDirectory);
    }

    private async Task WaitUntilHealthyAsync()
    {
        using HttpClient client = new();
        DateTimeOffset deadline = DateTimeOffset.UtcNow.AddSeconds(45);

        while (DateTimeOffset.UtcNow < deadline)
        {
            if (process is { HasExited: true })
            {
                throw new InvalidOperationException("nexus-server 在启动完成前退出。");
            }

            try
            {
                using HttpResponseMessage response = await client.GetAsync(runtime.HealthUrl);
                if (response.IsSuccessStatusCode)
                {
                    return;
                }
            }
            catch (HttpRequestException)
            {
                // sidecar 尚未监听端口，继续等待。
            }

            await Task.Delay(300);
        }

        throw new TimeoutException("等待 nexus-server 健康检查超时。");
    }
}

File: desktop/macos/Sources/NexusDesktop/Sidecar/SidecarSupervisor.swift
import Foundation

final class SidecarSupervisor {
  private let locator: SidecarBundleLocator
  private let port: Int
  private let runtimeConfig: SidecarRuntimeConfig
  private let orphanReaper: SidecarOrphanReaper
  private let startupTimeline: DesktopStartupTimeline?
  private let stdoutPipe = SidecarLogPipe(label: "stdout")
  private let stderrPipe = SidecarLogPipe(label: "stderr")
  private var process: Process?

  init(startupTimeline: DesktopStartupTimeline? = nil) throws {
    self.startupTimeline = startupTimeline
    locator = try SidecarBundleLocator.resolve()
    port = try SidecarPortAllocator.allocate()
    runtimeConfig = SidecarRuntimeConfig(port: port, sessionToken: try DesktopSessionToken.generate())
    orphanReaper = SidecarOrphanReaper(
      pidFileURL: DesktopPaths.sidecarPIDFileURL,
      expectedExecutablePath: locator.command
    )
    startupTimeline?.mark("sidecar.config_resolved", metadata: [
      "mode": locator.projectRoot == nil ? "bundle" : "development",
      "port": "\(port)",
    ])
  }

  func start() async throws -> SidecarRuntimeConfig {
    startupTimeline?.mark("sidecar.reap_begin")
    orphanReaper.reapIfNeeded()
    startupTimeline?.mark("sidecar.launch_begin")

    let sidecarProcess = Process()
    sidecarProcess.executableURL = URL(fileURLWithPath: locator.command)
    sidecarProcess.arguments = locator.arguments
    sidecarProcess.currentDirectoryURL = locator.workingDirectory
    sidecarProcess.environment = try buildEnvironment()
    sidecarProcess.standardOutput = stdoutPipe.fileHandleForWriting
    sidecarProcess.standardError = stderrPipe.fileHandleForWriting

    try sidecarProcess.run()
    process = sidecarProcess
    startupTimeline?.mark("sidecar.process_started", metadata: [
      "pid": "\(sidecarProcess.processIdentifier)",
    ])
    orphanReaper.write(pid: sidecarProcess.processIdentifier)
    do {
      try await waitUntilHealthy()
    } catch {
      stop()
      throw error
    }
    startupTimeline?.mark("sidecar.health_ready")
    return runtimeConfig
  }

  func stop() {
    defer {
      orphanReaper.removeRecord()
      stdoutPipe.close()
      stderrPipe.close()
    }

    guard let process, process.isRunning else {
      return
    }
    process.terminate()
    process.waitUntilExit()
  }

  private func buildEnvironment() throws -> [String: String] {
    var environment = ProcessInfo.processInfo.environment
    try DesktopPaths.createRuntimeDirectories()

    environment["NEXUS_APP_MODE"] = "desktop"
    environment["NEXUS_APP_ROOT"] = locator.appRootURL.path
    environment["NEXUS_CONFIG_DIR"] = DesktopPaths.configDirectory.path
    environment["HOST"] = "127.0.0.1"
    environment["PORT"] = "\(port)"
    environment["NEXUS_DESKTOP_SESSION_TOKEN"] = runtimeConfig.sessionToken
    environment["WEB_DIST_DIR"] = locator.webDistURL.path
    environment["DATABASE_DRIVER"] = "sqlite"
    environment["DATABASE_URL"] = DesktopPaths.dataDirectory.appendingPathComponent("nexus.db").path
    let credentialsKeyMode = connectorCredentialsKeyMode(environment: environment)
    startupTimeline?.mark("sidecar.credentials_key_begin", metadata: [
      "mode": credentialsKeyMode.rawValue,
    ])
    let credentialsKey = try DesktopKeychainStore.connectorCredentialsKey(mode: credentialsKeyMode)
    environment["CONNECTOR_CREDENTIALS_KEY"] = credentialsKey.value
    startupTimeline?.mark("sidecar.credentials_key_ready", metadata: [
      "mode": credentialsKeyMode.rawValue,
      "reason": credentialsKey.reason,
      "storage": credentialsKey.storage,
    ])
    environment["WORKSPACE_PATH"] = DesktopPaths.workspaceDirectory.path
    environment["CACHE_FILE_DIR"] = DesktopPaths.cacheDirectory.path
    environment["LOG_PATH"] = DesktopPaths.logsDirectory.appendingPathComponent("sidecar.log").path
    environment["LOG_STDOUT"] = "true"
    environment["LOG_FILE_ENABLED"] = "true"
    environment["DISCORD_ENABLED"] = "false"
    environment["TELEGRAM_ENABLED"] = "false"
    environment["CONNECTOR_OAUTH_REDIRECT_URI"] = "nexus://connectors/oauth/callback"
    applyPackagedConnectorConfig(to: &environment)
    let webOrigin = runtimeConfig.webURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    environment["CONNECTOR_OAUTH_ALLOWED_ORIGINS"] = "\(webOrigin),nexus://connectors"
    return environment
  }

  private func applyPackagedConnectorConfig(to environment: inout [String: String]) {
    let configURL = locator.appRootURL.appendingPathComponent("desktop.env")
    guard let data = try? String(contentsOf: configURL, encoding: .utf8) else {
      return
    }
    for rawLine in data.split(whereSeparator: \.isNewline) {
      let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
      if line.isEmpty || line.hasPrefix("#") {
        continue
      }
      let parts = line.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
      guard parts.count == 2 else {
        continue
      }
      let key = parts[0].trimmingCharacters(in: .whitespacesAndNewlines)
      let value = parts[1].trimmingCharacters(in: .whitespacesAndNewlines)
      if key.hasPrefix("CONNECTOR_") && !value.isEmpty {
        environment[key] = value
      }
    }
  }

  private func connectorCredentialsKeyMode(environment: [String: String]) -> DesktopKeychainMode {
    if let mode = DesktopKeychainMode(environmentValue: environment["NEXUS_DESKTOP_KEYCHAIN_MODE"]) {
      return mode
    }
    if locator.projectRoot != nil {
      return .file
    }
    return .auto
  }

  private func waitUntilHealthy() async throws {
    let deadline = Date().addingTimeInterval(45)
    while Date() < deadline {
      if let process, !process.isRunning {
        throw DesktopShellError.sidecarExited
      }
      if await isHealthy() {
        return
      }
      try await Task.sleep(nanoseconds: 300_000_000)
    }
    throw DesktopShellError.sidecarExited
  }

  private func isHealthy() async -> Bool {
    do {
      let (_, response) = try await URLSession.shared.data(from: runtimeConfig.healthURL)
      guard let httpResponse = response as? HTTPURLResponse else {
        return false
      }
      return (200..<300).contains(httpResponse.statusCode)
    } catch {
      return false
    }
  }

}
